### DISCUSSION TITLE
Replace image in PDF

### DISCUSSION THREAD

**Discussion by redkad** [pymupdf/PyMuPDF#936] [Looking for help] [answered]
Is there a way to replace images in PDF using PyMuPDF? I tried but I was only able to extract the images from the PDF.

**Comment by JorjMcKie** [comment]
duplicate of https://github.com/pymupdf/PyMuPDF/discussions/924.

**Reply by redkad** [reply]
Thanks a lot. It worked.

**Reply by redkad** [reply]
Please, Also, is there a way to replace a text with an image?

**Reply by JorjMcKie** [reply]
Yes, use the following steps:
1. Locate the text, e.g. via text search (`page.search_for("some text")`). Delivers a list of rectangles - text may occur multiple times.
2. Remove the text hits using redaction annotations.
3. Insert image(s) in the rectangles of the previously generated list.

Snippet:
```python
rect_list = page.search_for("some text")
for rect in rect_list:
    page.addRedactAnnot(rect, text=" ")
page.apply_redactions(images=fitz.PDF_REDACT_IMAGE_REMOVE)
# text occurrences removed now
# assuming, the same image goes into the emptied rectangles:
for rect in rect_list:
    page.insert_image(rect, pixmap=pix)
# instead of pixmap, an image filename can also be used
doc.save("newfile.pdf", garbage=4, deflate=True)
```

**Reply by redkad** [reply]
Ok. Thank You. I will try it out and you feedback. Thanks again.

**Reply by redkad** [reply]
It worked but the image stretched to fit the position of the text. Please is there a way to make the image fit well?

**Reply by redkad** [reply]
Please how can I get the pixmap?

**Reply by redkad** [reply]
So, I changed the pixmap=pix to filename="myimage.jpg". But I want to maintain the size of the image file. After it has been replaced by the text, the image is stretched to fill the position of the text which i don't want.

**Reply by JorjMcKie** [reply]
> Please how can I get the pixmap?

You mean the image you want to take as a text replacement?
It can be specified as a pixmap, image filename, or image binary content. So any of this:
```python
page.insert_image(rect, filename="image.file")
page.insert_image(rect, stream=<bytes/bytearray>)
page.insert_image(rect, pixmap=pix)
```
A pixmap can be created in many ways: ``pix = fitz.Pixmap("image.file")``, ``pix = page.get_pixmap(...)``, etc.

**Reply by JorjMcKie** [reply]
> But I want to maintain the size of the image file.

This is not possible: the image will always be scaled such that neither its height, nor its width will exceed the given rectangle

> the image is stretched to fill the position of the text

This can only happen if you specify `keep_proportion=False`, which is not the default. By default, the width-height-ratio is maintained.

**Comment by JorjMcKie** [comment]
> It worked but the image stretched to fit the position of the text. Please is there a way to make the image fit well?

The image will always be inserted such that its center and the center of the rectangle coincide. Then the image size is adjusted that its left-right or top-bottom sides coincide with the resp. rect sides.
Default is maintaining image aspect ratio. You can set `keep_proportion=False` and the image will stretch to cover the whole rectangle.

**Reply by redkad** [reply]
I have added the keep_proportion=False but nothing happened. What I actually want to achieve is to keep the image in a square position but not necessarily fitting into the rectangle of the text.

**Reply by JorjMcKie** [reply]
you **_must_** specify one of `keep_proportion=True` (default) or `False`. Whichever, the image will at least be scaled to one of redaction rect width or height.
Letting it default, the text rectangle should not be completely covered (if it doesn't happen to have the same propoertions).
If you see anything else, you are probably doing something wrong ...

**Reply by redkad** [reply]
Thank you very much for the information. Please I was wondering if there's a way to detect which portion of a PDF contains a text. For example. i need to know  if half the page contains  a text or quarter of the page contains a text. Thank you.

**Reply by JorjMcKie** [reply]
The higher text extractions contain bboxes for the text. Just add those bbox areas up (in a clever way!) and compare the result with the page's rect area.
"Clever" means: there may be overlaps between text bboxes which you may not want to count twice.
That has nothing to do with PyMuPDF - it is pure Python programming.

**Reply by redkad** [reply]
Thanks. Will give it a go. By the way do you have any code snippet to guide me?

**Comment by JorjMcKie** [comment, marked as answer]
> code snippet to guide me?

The rect area in PyMuPDF is coded as the builtin ``abs()`` function. It gives "square of points". There also are variants for square inches and centimeters, but for relative sizes, ``abs()`` is sufficient.
1. set `fitz.TOOLS.set_small_glyph_heights(True)`, to minimize the area each character occupies and any inter-span overlaps.
2. iterate over the text span bboxes and add there areas
3. then: `span_areas / abs(page.rect)`.

```python
>>> fitz.TOOLS.set_small_glyph_heights(True)
True
>>> blks=page.get_text("dict",flags=0)["blocks"]
>>> spans=[]
>>> for b in blks:
	for l in b["lines"]:
		for s in l["spans"]:
			spans.append(fitz.Rect(s["bbox"]))
>>> sum([abs(r) for r in spans]) / abs(page.rect)
0.23933628131017004
>>> 
```
So in this example, text should cover almost 24% of the page. In this case, there also were no span overlaps, so a precise result.
If you want to be awkwardly exact, add code to **_subtract_** areas of intersections of any two spans, and **_add_** again areas where three spans overlap, then **_subtract_** areas of intersections of 4 spans, ....

![grafik](https://user-images.githubusercontent.com/8290722/111587688-f29baf00-8798-11eb-9d54-86674ca6ddcc.png)

**Reply by redkad** [reply]
Thank you. Will try it out.
